### PR TITLE
Pile label

### PR DESCRIPTION
**Pull request by eeeeenchanted**
## Description

> What was changed in this pull request?

Add support for labels to piles via context menu.

> Why is it necessary?

## Checklist

- [x] GitHub labels properly set (e.g., bug, new feature, etc.)
- [x] Documentation added or updated
- [x] Examples added or updated
- [ ] Screenshot or GIF included (e.g. new or changed visualization features)
- [x] CHANGELOG.md updated

### CODE REVIEW

File: src/context-menu.js [hunk 1]
```diff
@@ -82,6 +82,15 @@ const TEMPLATE = `<ul id="piling-js-context-menu-list">
   <li>
     <button id="magnify-button">Magnify</button>
   </li>
+  <li>
+    <button id="label-green">Label: green</button>
```
Comment (flekschas) [inline src/context-menu.js]: Let's remove those predefined buttons. They can be added if needed by the user via custom context menu items.

File: src/pile.js [hunk 2]
```diff
@@ -1019,6 +1019,34 @@ const createPile = (
     setCover(newCover);
   };
 
+  const labelGraphics = new PIXI.Graphics();
```
Comment (flekschas) [inline src/pile.js]: Can we create the graphics object only if needed?

File: src/pile.js [hunk 3]
```diff
@@ -1019,6 +1019,34 @@ const createPile = (
...
+
+    const { width, height } = contentGraphics;
+
+    if (label.length) {
+      const labelWidth = width / label.length;
+      const labelHeight = 5;
+      label.forEach((color, index) => {
+        const labelX = labelWidth * index;
+        labelGraphics.beginFill(color, 1);
+        labelGraphics.drawRect(labelX, height, labelWidth, labelHeight);
```
Comment (flekschas) [inline src/pile.js]: To avoid the offset issues we need to subtract the half-width and half-height of the pile from x and y.

File: src/store.js [hunk 4]
```diff
@@ -488,17 +501,53 @@ const piles = (previousState = {}, action) => {
...
-            y: pile.y
+            y: pile.y,
+            label
           };
         });
       });
       return newState;
     }
 
+    case 'SET_LABEL': {
```
Comment (flekschas) [inline src/store.js]: Let's create a simple getter/setter property defining the name of the label property. Sorry I didn't specify this earlier but it'll simplify things quite a bit.

File: src/library.js [hunk 5]
```diff
@@ -2718,6 +2722,82 @@ const createPilingJs = (rootElement, initOptions = {}) => {
...
+    const color = label.map(l =>
+      isFunction(itemLabelColor)
+        ? itemLabelColor(l, labelArray)
+        : itemLabelColor[allLabels.get(l)]
+    );
+
+    let text = [];
+
+    if (itemLabelText) {
+      text = label.map(l => {
```
Comment (flekschas) [inline src/library.js]: For performance reasons, we should create unique `PIXI.Text` elements here and reuse them on the piles. Text elements are just another form of textures, so we should be able to stick them into `PIXI.Sprite` and position the sprite.

File: src/pile.js [hunk 6]
```diff
@@ -1032,6 +1032,54 @@ const createPile = (
...
+    if (isPositioning || isScaling) {
+      postPilePositionAnimation.set('drawLabel', () => {
+        drawLabel(labels, colors, texts);
+      });
+      return;
+    }
+
+    if (!labelGraphics || !labelTextContainer) {
+      labelGraphics = new PIXI.Graphics();
+      labelTextContainer = new PIXI.Container();
```
Comment (flekschas) [inline src/pile.js]: Do we really need another Container here? I think we can add the text directly to `labelGraphics`.

File: src/pile.js [hunk 7]
```diff
@@ -1032,6 +1032,54 @@ const createPile = (
...
+      contentGraphics.addChild(labelGraphics);
+      contentGraphics.addChild(labelTextContainer);
+    } else {
+      labelGraphics.clear();
+      labelTextContainer.removeChildren();
+    }
+
+    const { width, height } = contentGraphics;
+
+    if (labels.length) {
```
Comment (flekschas) [inline src/pile.js]: I think it'd be better to check directly at the beginning of the function if there are actually any labels to draw and if not, clear the graphics and exit the function. E.g.

```
if (!labels.length) {
  ... clear graphics
  return;
}
```

File: src/pile.js [hunk 8]
```diff
@@ -1032,6 +1032,54 @@ const createPile = (
...
+    }
+
+    if (texts.length) {
+      const textWidth = width / texts.length;
+      texts.forEach((text, index) => {
+        const labelText = new PIXI.Text(text);
+        labelText.anchor.set(0.5, 0);
+        labelText.x = textWidth * index - width / 2 + textWidth / 2;
+        labelText.y = height / 2;
+        labelText.style.fontSize = 15;
```
Comment (flekschas) [inline src/pile.js]: The font-size should ideally be user-controllable as well but at least it should be smaller by default :) Maybe 8px?

File: DOCS.md [hunk 9]
```diff
@@ -399,6 +402,49 @@ The list of all understood properties is given below.
...
+  piling.set('itemLabel', 'country');
+  piling.set('itemLabel', itemState => itemState.country);
+  piling.set('itemLabel', ['country', 'year']);
+  piling.set('itemLabel', { property: item => item.country, aggregator: countries => countries[0] });
+  ```
+
+- `itemLabelColor` can be set to an `array` of RGB strings, or a callback function. E.g.,
+
+  ```javascript
+  piling.set('itemLabelColor', ['0xe05aa9', '0xe0722b', '0xe0a638']);
```
[the remark this answers is not in the document]
Comment (flekschas) [reply DOCS.md]: Also, we should support other color definitions too. Basically we need to run the colors through `colorToDecAlpha()` before we set the color property. Take a look at the `pileBorderColor` setter. The alpha value isn't important for now.

File: src/library.js (hunk 5 above)
Comment (eeeeenchanted) [reply src/library.js]: Hmm I don't understand how to reuse the `PIXI.Text` element... If we stick it to `PIXI.Sprite`, then shouldn't we create multiple sprites?

Comment (flekschas) [reply src/library.js]: Can you take a look here? https://github.com/flekschas/piling.js/pull/138/files#diff-bab0fc1bfe78f6cc1e691c24be5a357cR21-R32

The idea is this:
1. Create a new `pixiText = new PIXI.Text('some text')` element for the label
2. Render the label using `pixiText.updateText()`
3. Extract texture using `texture = pixiText.texture`
4. Pass the texture to the pile, which can create a sprite out of it

This approach is muuuch more performant than create a new text element every time. I can explain it in more detail in our next meeting :)

File: src/pile.js [hunk 10]
```diff
@@ -1032,6 +1032,55 @@ const createPile = (
...
+      labelGraphics.clear();
+      labelGraphics.removeChildren();
+    }
+
+    const { width, height } = contentGraphics;
+
+    const labelWidth = width / labels.length;
+    const labelHeight = 8;
+    labels.forEach((label, index) => {
+      const labelX = labelWidth * index - width / 2;
```
Comment (flekschas) [inline src/pile.js]: I think this position needs to be adjusted with on the `baseScale`. Maybe `labelX /= baseScale`?

**Comment by flekschas** [comment]
I realized we need a few more essential properties for the labels: `pileLabelHeight`, `pileLabelFontSize`, `pileLabelAlign`, and `pileLabelStackAlign`:
- `pileLabelHeight` should simply control `labelHeight`. When text is displayed too we might have to set `pileLabelHeight = Math.max(pileLabelHeight, pileLabelFontSize)`
- `pileLabelFontSize` should control the text font size
- `pileLabelAlign` should define the alignment, which we can restrict to `top` and `bottom` for now. By default, let's set it to `bottom`
- `pileLabelStackAlign` should define the alignment of multiple labels. We can limit it to `horizontal` and `vertical` for now and set the default to `horizontal` (which is what we have right now).

Can you implement those things as well? :)

**Pull request merged**